Adds documentation and example recommending Vec<ArrayRef> over ChunkedArray

efredine opened the pull request:
# Which issue does this PR close?

Closes #5295.

# Rationale for this change
 
As described in issue.

# What changes are included in this PR?

Adds documentation and an example

# Are there any user-facing changes?

Only documentation and example.

tustvold commented on arrow/src/lib.rs:
```diff
@@ -345,6 +345,67 @@
+//! This approach integrates well into the Rust ecosystem, simplifies the implementation and
+//! encourages the use of performant lazy and async patterns.
+//!
+//! Aside from providing a slightly less convenient API, one other downside is the lack of support
+//! for processing compute kernels across chunked arrays. But this use case is well-supported by
+//! [DataFusion].
+//!
```
```suggestion
```

I'm not sure I agree with this statement, iterators are fairly ergonomic once you get the hang of them

tustvold commented on arrow/examples/chunked_arrays.rs:
```diff
@@ -0,0 +1,106 @@
+        .unwrap(),
+    ];
+
+    // chunked_array_by_index is an array of two Vec<ArrayRef> where each Vec<ArrayRef> is a column
+    let mut chunked_array_by_index = [Vec::new(), Vec::new()];
+    for batch in &batches {
+        for (i, array) in batch.columns().iter().enumerate() {
+            chunked_array_by_index[i].push(array.clone());
+        }
+    }
```
Collecting into separate Vec like this is unnecessary, see below

tustvold commented on arrow/examples/chunked_arrays.rs:
```diff
@@ -0,0 +1,106 @@
+    // downcast and iterate over the values - column 0 is the labels and column 1 is the values
+    let labels: Vec<&str> = chunked_array_by_index[0]
+        .iter()
+        .flat_map(|x| as_string_array(x).iter())
```
Using AsArray is typically more ergonomic

tustvold reviewed (commented):
I think this example could be drastically simplified by using the iterators fully, e.g.

```
let values: Vec<f32> = batches.iter().flat_map(|x| x.column(0).as_primitive::<Float32Type>().values()).collect();

let values: Vec<&str> = batches.iter().flat_map(|x| x.column(1).as_string()).map(Option::unwrap).collect();
```

I can't see an obvious reason to collect into separate Vecs, this just adds additional overheads, and is less concise.

I think it is also worth highlighting that many use-cases won't need to cast down from the ArrayRef at all, with most kernels accept `&dyn Array`.

tustvold commented on arrow/src/lib.rs:
```diff
@@ -345,6 +345,67 @@
 //! orchestrates the primitives exported by this crate into an embeddable query engine, with
 //! SQL and DataFrame frontends, and heavily influences this crate's roadmap.
 //!
+//! The Rust implementation does not provide the ChunkedArray abstraction implemented by the Python
```
I think this documentation probably belongs at the end of these docs - https://docs.rs/arrow-array/latest/arrow_array/

Not the top-level crate documentation

efredine commented:
Thanks @tustvold - I've updated the text, simplified the example as suggested and moved it to arrow_array. (I also deleted the superfluous example I had created).

efredine commented on arrow/examples/README.md:
```diff
@@ -21,7 +21,7 @@
 
 - [`builders.rs`](builders.rs): Using the Builder API
 - [`collect.rs`](collect.rs): Using the `FromIter` API
-- [`dynamic_types.rs`](dynamic_types.rs):
```
Unrelated change - this example's description was missing from the README.

tustvold reviewed (approved):
Looks good to me, thank you

tustvold commented on arrow-array/src/lib.rs:
```diff
@@ -161,7 +161,55 @@
+//!
+//! This approach integrates well into the Rust ecosystem, simplifies the implementation and
+//! encourages the use of performant lazy and async patterns.
+//!
+//! The iterator API makes it ergonomic to work with these patterns and since most kernels accept
+//! `&dyn Array` many use-cases do not even require downcasting.
```
```suggestion
```
I think this is probably redundant with the section on downcasting above, but I don't fell strongly
